include/merlin_hashtable.cuh

```diff
@@ -431,6 +453,8 @@ class HashTable {
                         score_type* evicted_scores,    // (n)
                         size_type* d_evicted_counter,  // (1)
                         cudaStream_t stream = 0,
+                        const score_type global_epoch =
```

Comment:
Hi @Lifann , please help review the change on API, this PR implemented the new strategy which needs `global_epoch` as the high 32bit of the score, for more detail please refer to https://github.com/rhdong/HierarchicalKV/tree/rhdong/epoch-lru#evict-strategy

Reply:
A fixed data format param for limited case will make the API hard to expand in the future.

Reply:
I agree with you, but it seems no better choice. Do you have a better recommendation?

Reply:
Here is a possible choice:
```cpp
void global_epoch(const S epoch);
S global_epoch() const;
```

and add a `std::atomic<S> global_epoch;` member for the Hashtable.

Reply:
```cpp
/*
 * A prototype to show how to get param inside functor implicitly.
 */

#include "cuda_runtime.h"
#include <stdio.h>
#include <vector>
#include <unistd.h>
using namespace std;

namespace functor {

__device__ int param_;

void set_param(int score) {
  int ret = 0;
  ret = cudaMemcpyToSymbol(param_, &score, sizeof(int), 0, cudaMemcpyHostToDevice);
  printf("symbol op1 ret=%d\n", ret);
}

struct Functor {
 public:
  __device__ inline int internal_get_param() {
    return ::functor::param_;
  }
  __device__ void operator()(int& key, int& score) {
    int inc = internal_get_param();
    score = inc;
  }
};

}

template <typename ScorePred>
__global__ void gpu_any_kernel(int* keys, int* scores, size_t n) {
  int tid = blockDim.x * blockIdx.x + threadIdx.x;
  ScorePred pred;
  if (tid < n) {
    pred(keys[tid], scores[tid]);
  }
}

void get_data(vector<int>& data, int** d_data) {
  int ret = 0;
  cudaMalloc(d_data, data.size() * sizeof(int));
  cudaMemcpy(*d_data, data.data(), data.size() * sizeof(int), cudaMemcpyHostToDevice);
}

int main() {
  int ret = 0;
  cudaSetDevice(0);
  cudaStream_t stream;
  cudaStreamCreate(&stream);
  vector<int> keys = {1,2,3,4,5};
  vector<int> scores = {0,0,0,0,0};
  int* d_keys = nullptr;
  int* d_scores = nullptr;
  get_data(keys, &d_keys);
  get_data(scores, &d_scores);
  functor::set_param(5);
  cudaDeviceSynchronize();

  // check whether if set_param work.
  gpu_any_kernel<functor::Functor><<<5, 1, 0, stream>>>(d_keys, d_scores, keys.size());
  ret = cudaMemcpyAsync(scores.data(), d_scores, sizeof(int) * scores.size(), cudaMemcpyDeviceToHost, stream);
  printf("copy d2h ret=%d\n", ret);
  cudaStreamSynchronize(stream);
  for (int i = 0; i < 5; i++) {
    printf("i=%d, score=%d\n", i, scores[i]); // It will get 5 on all pos.
  }
  return 0;
}
```

Reply:
> ```c++
> /*
>  * A prototype to show how to get param inside functor implicitly.
>  */
> 
> #include "cuda_runtime.h"
> #include <stdio.h>
> #include <vector>
> #include <unistd.h>
> using namespace std;
> 
> namespace functor {
> 
> __device__ int param_;
> 
> void set_param(int score) {
>   int ret = 0;
>   ret = cudaMemcpyToSymbol(param_, &score, sizeof(int), 0, cudaMemcpyHostToDevice);
>   printf("symbol op1 ret=%d\n", ret);
> }
> 
> struct Functor {
>  public:
>   __device__ inline int internal_get_param() {
>     return ::functor::param_;
>   }
>   __device__ void operator()(int& key, int& score) {
>     int inc = internal_get_param();
>     score = inc;
>   }
> };
> 
> }
> 
> template <typename ScorePred>
> __global__ void gpu_any_kernel(int* keys, int* scores, size_t n) {
>   int tid = blockDim.x * blockIdx.x + threadIdx.x;
>   ScorePred pred;
>   if (tid < n) {
>     pred(keys[tid], scores[tid]);
>   }
> }
> 
> void get_data(vector<int>& data, int** d_data) {
>   int ret = 0;
>   cudaMalloc(d_data, data.size() * sizeof(int));
>   cudaMemcpy(*d_data, data.data(), data.size() * sizeof(int), cudaMemcpyHostToDevice);
> }
> 
> int main() {
>   int ret = 0;
>   cudaSetDevice(0);
>   cudaStream_t stream;
>   cudaStreamCreate(&stream);
>   vector<int> keys = {1,2,3,4,5};
>   vector<int> scores = {0,0,0,0,0};
>   int* d_keys = nullptr;
>   int* d_scores = nullptr;
>   get_data(keys, &d_keys);
>   get_data(scores, &d_scores);
>   functor::set_param(5);
>   cudaDeviceSynchronize();
> 
>   // check whether if set_param work.
>   gpu_any_kernel<functor::Functor><<<5, 1, 0, stream>>>(d_keys, d_scores, keys.size());
>   ret = cudaMemcpyAsync(scores.data(), d_scores, sizeof(int) * scores.size(), cudaMemcpyDeviceToHost, stream);
>   printf("copy d2h ret=%d\n", ret);
>   cudaStreamSynchronize(stream);
>   for (int i = 0; i < 5; i++) {
>     printf("i=%d, score=%d\n", i, scores[i]); // It will get 5 on all pos.
>   }
>   return 0;
> }
> ```

In the case, it's possible to apply any param implicitly on functor, which can traverse on the key-value-score pairs of inputs of function call.

Reply:
I merged our thoughts here, please help review again, thanks: https://github.com/NVIDIA-Merlin/HierarchicalKV/pull/152/commits/afd5813706052b9b27101f6a309c1344cc836698